src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,157 @@
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     *
+     * @since 2.16
+     */
+    class NonRecyclingRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ConcurrentDequePool implements BufferRecyclerPool {
+        private final Deque<BufferRecycler> pool = new ConcurrentLinkedDeque<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
+        }
+
+        private BufferRecycler getBufferRecycler() {
+            BufferRecycler bufferRecycler = pool.pollFirst();
+            return bufferRecycler != null ? bufferRecycler : new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+            pool.offerLast(bufferRecycler);
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * a lock free linked list for recycling instances.
+     *
+     * @since 2.16
+     */
+    class LockFreePool implements BufferRecyclerPool {
+        private final AtomicReference<LockFreePool.Node<BufferRecycler>> head = new AtomicReference<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
+        }
+
+        private BufferRecycler getBufferRecycler() {
+            for (int i = 0; i < 3; i++) {
+                Node<BufferRecycler> currentHead = head.get();
+                if (currentHead == null) {
+                    return new BufferRecycler();
+                }
+                if (head.compareAndSet(currentHead, currentHead.next)) {
+                    currentHead.next = null;
+                    return currentHead.value;
+                }
+            }
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+            LockFreePool.Node<BufferRecycler> newHead = new LockFreePool.Node<>(bufferRecycler);
+            for (int i = 0; i < 3; i++) {
+                newHead.next = head.get();
+                if (head.compareAndSet(newHead.next, newHead)) {
+                    return;
+                }
+            }
+        }
+
+        static class Node<T> {
```

Comment:
Is there point to making this generic, if only used with `Node<BufferRecycler>` parameter?

Reply:
Good point, I just copied a reference implementation. I will remove the generic here.